### PR TITLE
Make `Rand` deterministic

### PR DESCRIPTION
**Pull request by geoffromer**
Also re-enable auto-updating of `Rand` test (disabled in #1894)

### CODE REVIEW

**Comment by SlaterLatiao** [comment]
Please correct me if I'm wrong. I think `Rand` uses `std::uniform_int_distribution` which does not guarantee the same results across platforms.

**Review by SlaterLatiao — CHANGES_REQUESTED** [review]
I think the result of `std::uniform_int_distribution` is only deterministic on a given platform and standard library implementation.

**Comment by geoffromer** [comment]
> Please correct me if I'm wrong. I think `Rand` uses `std::uniform_int_distribution` which does not guarantee the same results across platforms.

Good point. OK, I've removed the dependency on `std::uniform_int_distribution`.

**Review by SlaterLatiao — APPROVED** [review]
Looking good!

File: explorer/interpreter/interpreter.cpp [hunk 1]
```diff
@@ -1209,8 +1209,9 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
           CARBON_CHECK(args.size() == 2);
           const auto& low = cast<IntValue>(*args[0]).value();
           const auto& high = cast<IntValue>(*args[1]).value();
-          std::uniform_int_distribution<> distr(low, high);
-          int r = distr(generator);
+          // We avoid using std::uniform_int_distribution because it's not
+          // reproducible across builds/platforms.
+          int r = (generator() % (low - high)) + low;
```
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]: Wouldn't you normally do `high - low`?

Should this check that high > low?

Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp]: If I called the prior implementation, or std::uniform_int_distribution, with Rand(0,1) I believe the result is [0, 1] (inclusive). What you've written here would be [0, 1). Thoughts?

How important is keeping a random number generator, for the time we're spending on this? Is this advancing the language goals?

Comment (zygoloid) [reply explorer/interpreter/interpreter.cpp]: Also, how important is it that this is an intrinsic? I would think we can implement a simple PRNG in the prelude if we really need one in the short term.

Comment (geoffromer) [reply explorer/interpreter/interpreter.cpp]: > Wouldn't you normally do `high - low`?

Whoops. Fixed.

> Should this check that high > low?

Sure, done.

> How important is keeping a random number generator, for the time we're spending on this? Is this advancing the language goals?

> Also, how important is it that this is an intrinsic? I would think we can implement a simple PRNG in the prelude if we really need one in the short term.

I think `Rand` in itself is not very important, but I think it's valuable to have an example of how to implement a bog-standard intrinsic, now that `Print` can't fill that role. So I'd prefer to keep `Rand`, and keep it as an intrinsic, for now.

> If I called the prior implementation, or std::uniform_int_distribution, with Rand(0,1) I believe the result is [0, 1] (inclusive). What you've written here would be [0, 1). Thoughts?

I think that's fine; it doesn't need to be a good PRNG, it just needs to be roughly PRNG-shaped, deterministic to keep testing simple, and implemented as an intrinsic.

Comment (zygoloid) [reply explorer/interpreter/interpreter.cpp]: Sorry for the potential side-track, but maybe we can implement `Print` via standard intrinsics:
```
interface Printable {
  fn Print[me: Self]();
}
impl i32 as Printable {
  fn Print[me: Self]() { __intrinsic_int_print(me); }
}
impl String as Printable {
  fn Print[me: Self]() { __intrinsic_string_print(me); }
}
fn Print[T:! Printable](x: T) { x.Print(); }
```

Comment (jonmeow) [reply explorer/interpreter/interpreter.cpp]: Would that work for `Print(format, val)`? (I don't think it can?)

**Review by jonmeow — APPROVED** [review]
I'm going to approve this, but it may still be worth thinking about other standard functions with more reliable behavior if you want a wrapper (exit/abort?)

**Pull request merged**